### PR TITLE
Refactor config

### PR DESCRIPTION
**Pull request by Lee-W**
## Description 
* refactor(conventional_commits): remove duplicate patterns and import from defaults
* refactor(defaults):
    * add CzSettings and Questions TypedDict
    * add Settings typeddict
    * move bump_map, bump_pattern, commit_parser from defaults to ConventionalCommitsCz
    * import TypedDict from type_extensions for backward compatibility
    * annoate OrderedDict through 'OrderedDict' for 3.6 compatibility
* style
    * enforce warn_return_any check
    * disable disallow_untyped_decorators for tests
    * add return type to all functions
* test: move default config to top level conftest.py
* style: format code base with latest black
* build(poetry): update black version for python 3.9 and upgrade least support python version to python 3.6.2

## Checklist

- [x] Add test cases to all the changes you introduce
- [x] Run `./script/format` and `./script/test` locally to ensure this change passes linter check and test
- [x] Test the changes on the local machine manually
- [x] Update the documentation for the changes

## Expected behavior
no new features has been added. all the functionality should be the same as it used to be

## Additional context
#424, #153

### CODE REVIEW

**Comment by Lee-W** [comment]
@Woile this branch will need to wait for https://github.com/commitizen-tools/commitizen/pull/456 to be merged. Although this is ready to be reviewed, I'm thinking of adding the functionality of validating config (https://github.com/commitizen-tools/commitizen/issues/300) and overriding customize config in base config (related to https://github.com/commitizen-tools/commitizen/issues/264#issuecomment-891453998). do you think I should finish all those features then send them to you? or do you want me to create separate PRs for those functionalities?

**Comment by Woile** [comment]
This looks good, I would go one step at a time. 
I would definitely update the documentation, specially to mention the new data types when building a custom commitizen ruler.
Using `questions(...) -> list:` is quite intuitive because everyone knows what a `list` is. If using `Questions` it should be documented and explained that it is basically the same as a list, what do you think?

**Comment by Lee-W** [comment]
One of the goals of this and the following PRs is to remove Generic type annotation and go for specific ones. So I would prefer to use either `Question` or `Iterable[MutableMapping[str, Any]]`. I'll add the documentation to address it these days. Will change this branch from draft to ready when I've done so.

**Comment by Woile** [comment]
Sounds good to me 💪🏻 looking forward

**Comment by Lee-W** [comment]
I've updated the content in `docs/customization.md`  Could you please check whether I need to add it to elsewhere? Thanks!

File: docs/customization.md [hunk 1]
```diff
@@ -194,15 +194,15 @@ See [commitizen_cz_template](https://github.com/commitizen-tools/commitizen_cz_t
...
+Inherit from `BaseCommitizen`, and you must define `questions` and `message`. The others are optional.
 
 ```python
 from commitizen.cz.base import BaseCommitizen
+from commitizen.defaults import Questions
 
 class JiraCz(BaseCommitizen):
-
-    def questions(self) -> list:
+    # Questions = Iterable[MutableMapping[str, Any]]
```
Comment (Woile) [inline docs/customization.md:204]: I'd also explain it in simpler terms:

> Questions expect a list with dictionaries

File: commitizen/defaults.py [hunk 2]
```diff
@@ -1,5 +1,41 @@
...
+    questions: Questions
+    example: Optional[str]
+    schema_pattern: Optional[str]
+    schema: Optional[str]
+    info_path: Union[str, pathlib.Path]
+    info: str
+    message_template: str
+
+
+class Settings(TypedDict, total=False):
```
Comment (Woile) [inline commitizen/defaults.py:25]: Beautiful!

**Review by Woile — APPROVED** [review]

File: docs/customization.md (hunk 1 above)
Comment (Lee-W) [reply docs/customization.md:204]: sounds good. just update it

**Comment by Woile** [comment]
I think is good. The commit has a typo only 😅

**Comment by Lee-W** [comment]
Do you mean the missing s after expect? If so, I've fixed that. If not so, could please you point it out 😱

**Comment by Lee-W** [comment]
As this is approved, and I tried to address the typo. I plan to merge this one late today
